src/commands/watch.rs

```diff
@@ -0,0 +1,304 @@
+    handler::SyncFnHandler,
+    signal::source::MainSignal::Interrupt,
+    ErrorHook, Watchexec,
+};
+
+use crate::{
+    opts::{
+        APP_MANIFEST_FILE_OPT, DEFAULT_MANIFEST_FILE, WATCH_CLEAR_OPT, WATCH_DEBOUNCE_OPT,
+        WATCH_SKIP_BUILD_OPT, WATCH_WATCH_ASSETS_OPT,
+    },
+    watch_filter::{WatchFilter, DEFAULT_IGNORE_PATTERNS},
+};
+
+// TODO: Add a --component <COMPONENT> flag
+// TODO: Add a --watch <WATCH_CONFIG> flag
+// TODO: Add a --ignore <IGNORE_CONFIG> flag
+
+/// Execute spin build and spin up when watched files change.
+#[derive(Parser, Debug)]
+#[clap(
+    about = "Rebuild and restart the Spin application when files changes",
+    allow_hyphen_values = true
+)]
+pub struct WatchCommand {
+    /// Path to spin.toml.
+    #[clap(
+            name = APP_MANIFEST_FILE_OPT,
+            short = 'f',
+            long = "file",
+            default_value = DEFAULT_MANIFEST_FILE,
+        )]
+    pub app: PathBuf,
+
+    /// Clear the screen before each run.
+    #[clap(
+            name = WATCH_CLEAR_OPT,
+            short = 'c',
+            long = "clear",
+    )]
+    pub clear: bool,
+
+    /// Set the timeout between detected change and re-execution.
+    #[clap(
+        name = WATCH_DEBOUNCE_OPT,
+        short = 'd',
+        long = "debounce",
+        default_value = "100"
+    )]
+    pub debounce: u64,
+
+    /// Skip running spin build and only run spin up.
+    #[clap(name = WATCH_SKIP_BUILD_OPT, long = "skip-build")]
+    pub skip_build: bool,
+
+    /// Only watch component source and files ignoring watch configuration in spin.toml.
+    #[clap(name = WATCH_WATCH_ASSETS_OPT, long = "watch-assets", requires = WATCH_SKIP_BUILD_OPT)]
+    pub watch_assets: bool,
+
+    /// Arguments to be passed through to spin up.
+    #[clap()]
+    pub up_args: Vec<OsString>,
+}
+
+impl WatchCommand {
+    pub async fn run(self) -> Result<()> {
+        // Prepare InitConfig for Watchexec
+        let mut init_config = InitConfig::default();
+        init_config.on_error(SyncFnHandler::from(
+            |err: ErrorHook| -> std::result::Result<(), Infallible> {
+                // This is a spurious error that we don't want to log
+                if let RuntimeError::IoError {
+                    about: "waiting on process group",
+                    ..
+                } = err.error
+                {
+                    return Ok(());
+                }
+
+                tracing::error!("{}", err.error);
+                Ok(())
+            },
+        ));
+
+        // Prepare RuntimeConfig for Watchexec
+        let mut runtime_config = RuntimeConfig::default();
+        let (spin_cmd, spin_args) = self.generate_command();
+        runtime_config.commands(vec![watchexec::command::Command::Exec {
+            prog: spin_cmd,
+            args: spin_args,
+        }]);
+        let app_dir = parent_dir(&self.app)?;
+        runtime_config.pathset([app_dir.to_str().unwrap().to_owned()]);
```

Comment:
Based on [the docs](https://docs.rs/watchexec/2.3.0/watchexec/config/struct.RuntimeConfig.html#method.pathset) it doesn't look like you need to convert to a string here. Maybe this would work:
```suggestion
        runtime_config.pathset([app_dir.clone()]);
```

Reply:
Worked like a charm.